syntax = "proto3";

package reddit.comment;

import "user.proto";

// Objects
message Comment {
  optional user.User author = 1;
  optional int32 score = 2;
  optional CommentState state = 3;
  optional string published = 4;
  optional string content = 5;
  optional int64 ID = 6;
  repeated Comment comment = 7;

  enum CommentState {
    NORMAL = 0;
    HIDDEN = 1;
  }
}

// Requests/responses
message NewCommentRequest {
  optional user.User author = 1;
  optional Comment.CommentState state = 2;
  optional string content = 3;
}

message NewCommentResponse {
  optional Comment comment = 1;
}

message RateCommentRequest {
  optional int64 commentID = 1;
  optional string rating = 2;

  enum Rating {
    UPVOTE = 0;
    DOWNVOTE = 1;
  }
}

message RateCommentResponse {
  optional Comment comment = 1;
}

message GetCommentRequest{
  optional int64 commentID = 1;
}

message GetCommentResponse{
  optional Comment comment = 1;
}

message GetNCommentsRequest {
  optional int32 num_comments = 1;
  optional int64 commentID = 2;
}

message GetNCommentsResponse {
  optional Comment comment = 1;
  optional Comment subcomment = 2;
}

//services
service CommentService {
  // send an author, text, and initial state and return the new object
  rpc CreateComment (NewCommentRequest) returns (NewCommentResponse) {}
  // send a comment ID and a rating, get the new rating back
  rpc RateComment (RateCommentRequest) returns (RateCommentResponse) {}
  // send a post id, and get its metadata back
  rpc GetComment (GetCommentRequest) returns (GetCommentResponse) {}
  // send comment ID and number of comments to get from it and get N most upvoted comments underneath
  rpc GetNComments (GetNCommentsRequest) returns (stream GetNCommentsResponse) {}
}